ports/c#/test/Antlr4CodeCompletion.CoreUnitTest/CodeCompletion/CodeCompletionCoreUnitTests.cs

```diff
@@ -0,0 +1,205 @@
+
+            // 5) On the variable reference 'a'. But since we have not configure the c3 engine to return us var refs
+            //    (or function refs for that matter) we only get an ID here.
+            candidates = core.CollectCandidates(6, null);
+            Check.That(candidates.Tokens).HasSize(1);
+            Check.That(candidates.Tokens).ContainsKey(ExprLexer.ID);
+
+            // 6) On the '+' operator. Usually you would not show operators as candidates, but we have not set up the c3 engine
+            //    yet to not return them.
+            candidates = core.CollectCandidates(8, null);
+            Check.That(candidates.Tokens).HasSize(5);
+            Check.That(candidates.Tokens).ContainsKey(ExprLexer.PLUS);
+            Check.That(candidates.Tokens).ContainsKey(ExprLexer.MINUS);
+            Check.That(candidates.Tokens).ContainsKey(ExprLexer.MULTIPLY);
+            Check.That(candidates.Tokens).ContainsKey(ExprLexer.DIVIDE);
+            Check.That(candidates.Tokens).ContainsKey(ExprLexer.OPEN_PAR);
+        }
+
+        [Fact]
+        public void Completion_Grammar_TypicalExpression()
+        {
+            // arrange
+            var expression = "var c = a + b";
+            var inputStream = new AntlrInputStream(expression);
+            var lexer = new ExprLexer(inputStream);
+            var tokenStream = new CommonTokenStream(lexer);
+            var parser = new ExprParser(tokenStream);
+            parser.Interpreter.PredictionMode = PredictionMode.LlExactAmbigDetection;
+
+            lexer.RemoveErrorListeners();
+            parser.RemoveErrorListeners();
+
+            var errorListener = new CountingErrorListener();
+            parser.AddErrorListener(errorListener);
+
+            // act
+            // assert
+
+            // Specify our entry point
+            var tree = parser.expression();
+
+            Check.That(errorListener.ErrorCount).IsEqualTo(0);
+
+            // Tell the engine to return certain rules to us, which we could use to look up values in a symbol table.
+            var preferredRules = new HashSet<int>() { ExprParser.RULE_functionRef, ExprParser.RULE_variableRef };
+
+            // Ignore operators and the generic ID token.
+            var ignoredTokens = new HashSet<int>() { ExprLexer.ID, ExprLexer.PLUS, ExprLexer.MINUS,
+                ExprLexer.MULTIPLY, ExprLexer.DIVIDE, ExprLexer.EQUAL };
+
+            var core = new CodeCompletionCore(parser, preferredRules, ignoredTokens);
+
+            // 1) At the input start.
+            var candidates = core.CollectCandidates(0, null);
+
+            Check.That(candidates.Tokens).HasSize(2);
+            Check.That(candidates.Tokens).ContainsKey(ExprLexer.VAR);
+            Check.That(candidates.Tokens).ContainsKey(ExprLexer.LET);
+
+            Check.That(candidates.Tokens[ExprLexer.VAR]).HasSize(2);
+            Check.That(candidates.Tokens[ExprLexer.LET]).HasSize(2);
```

Comment:
I neither see these two checks in the TS version nor in Java. However, the check is ok, as you are testing that there are 2 additional tokens stored for `var` and `let`. However, when line 135 fails then line 138 should fail too, not line 139.

It's hard to say for me why for the VAR token there are not 2 additional tokens stored. You can only find out by debugging the collection process.